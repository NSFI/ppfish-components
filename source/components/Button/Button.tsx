import * as React from 'react';
import { findDOMNode } from 'react-dom';
import * as PropTypes from 'prop-types';
import classNames from 'classnames';
import Icon from '../Icon';
import Group from './ButtonGroup';

const rxTwoCNChar = /^[\u4e00-\u9fa5]{2}$/;
const isTwoCNChar = rxTwoCNChar.test.bind(rxTwoCNChar);
function isString(str: any) {
  return typeof str === 'string';
}

// Insert one space between two chinese characters automatically.
function insertSpace(child: React.ReactChild, needInserted: boolean) {
  // Check the child if is undefined or null.
  if (child == null) {
    return;
  }
  const SPACE = needInserted ? ' ' : '';
  // strictNullChecks oops.
  if (typeof child !== 'string' && typeof child !== 'number' &&
    isString(child.type) && isTwoCNChar(child.props.children)) {
    return React.cloneElement(child, {},
      child.props.children.split('').join(SPACE));
  }
  if (typeof child === 'string') {
    if (isTwoCNChar(child)) {
      child = child.split('').join(SPACE);
    }
    return <span>{child}</span>;
  }
  return child;
}

export type ButtonType = 'default' | 'primary' | 'ghost' | 'dashed' | 'danger';
export type ButtonShape = 'circle' | 'circle-outline';
export type ButtonSize = 'small' | 'default' | 'large';
export type ButtonHTMLType = 'submit' | 'button' | 'reset';

export interface BaseButtonProps {
  type?: ButtonType;
  icon?: string;
  shape?: ButtonShape;
  size?: ButtonSize;
  loading?: boolean | { delay?: number };
  prefixCls?: string;
  className?: string;
  ghost?: boolean;
}

export type AnchorButtonProps = {
  href: string;
  target?: string;
  onClick?: React.MouseEventHandler<HTMLAnchorElement>;
} & BaseButtonProps & React.AnchorHTMLAttributes<HTMLAnchorElement>;

export type NativeButtonProps = {
  htmlType?: ButtonHTMLType;
  onClick?: React.MouseEventHandler<HTMLButtonElement>;
} & BaseButtonProps & React.ButtonHTMLAttributes<HTMLButtonElement>;

export type ButtonProps = AnchorButtonProps | NativeButtonProps;

export default class Button extends React.Component<ButtonProps, any> {
  static Group: typeof Group;
  static __ANT_BUTTON = true;

  static defaultProps = {
    prefixCls: 'ant-btn',
    loading: false,
    ghost: false,
  };

  static propTypes = {
    type: PropTypes.string,
    shape: PropTypes.oneOf(['circle', 'circle-outline']),
    size: PropTypes.oneOf(['large', 'default', 'small']),
    htmlType: PropTypes.oneOf(['submit', 'button', 'reset']),
    onClick: PropTypes.func,
    loading: PropTypes.oneOfType([PropTypes.bool, PropTypes.object]),
    className: PropTypes.string,
    icon: PropTypes.string,
  };

  timeout: number;
  delayTimeout: number;

  constructor(props: ButtonProps) {
    super(props);
    this.state = {
      loading: props.loading,
      clicked: false,
      hasTwoCNChar: false,
    };
  }

  componentDidMount() {
    this.fixTwoCNChar();
  }

  componentWillReceiveProps(nextProps: ButtonProps) {
    const currentLoading = this.props.loading;
    const loading = nextProps.loading;

    if (currentLoading) {
      clearTimeout(this.delayTimeout);
    }

    if (typeof loading !== 'boolean' && loading && loading.delay) {
      this.delayTimeout = window.setTimeout(() => this.setState({ loading }), loading.delay);
    } else {
      this.setState({ loading });
    }
  }

  componentDidUpdate() {
    this.fixTwoCNChar();
  }

  componentWillUnmount() {
    if (this.timeout) {
      clearTimeout(this.timeout);
    }
    if (this.delayTimeout) {
      clearTimeout(this.delayTimeout);
    }
  }

  fixTwoCNChar() {
    // Fix for HOC usage like <FormatMessage />
    const node = (findDOMNode(this) as HTMLElement);
    const buttonText = node.textContent || node.innerText;
    if (this.isNeedInserted() && isTwoCNChar(buttonText)) {
      if (!this.state.hasTwoCNChar) {
        this.setState({
          hasTwoCNChar: true,
        });
      }
    } else if (this.state.hasTwoCNChar) {
      this.setState({
        hasTwoCNChar: false,
      });
    }
  }

  handleClick: React.MouseEventHandler<HTMLButtonElement | HTMLAnchorElement> = e => {
    // Add click effect
    this.setState({ clicked: true });
    clearTimeout(this.timeout);
    this.timeout = window.setTimeout(() => this.setState({ clicked: false }), 500);

    const onClick = this.props.onClick;
    if (onClick) {
      (onClick as React.MouseEventHandler<HTMLButtonElement | HTMLAnchorElement>)(e);
    }
  }

  isNeedInserted() {
    const { icon, children } = this.props;
    return React.Children.count(children) === 1 && !icon;
  }

  render() {
    const {
      type, shape, size, className, children, icon, prefixCls, ghost, loading: _loadingProp, ...rest
    } = this.props;

    const { loading, clicked, hasTwoCNChar } = this.state;

    // large => lg
    // small => sm
    let sizeCls = '';
    switch (size) {
      case 'large':
        sizeCls = 'lg';
        break;
      case 'small':
        sizeCls = 'sm';
      default:
        break;
    }

    const classes = classNames(prefixCls, className, {
      [`${prefixCls}-${type}`]: type,
      [`${prefixCls}-${shape}`]: shape,
      [`${prefixCls}-${sizeCls}`]: sizeCls,
      [`${prefixCls}-icon-only`]: !children && icon,
      [`${prefixCls}-loading`]: loading,
      [`${prefixCls}-clicked`]: clicked,
      [`${prefixCls}-background-ghost`]: ghost,
      [`${prefixCls}-two-chinese-chars`]: hasTwoCNChar,
    });

    const iconType = loading ? 'loading' : icon;
    const iconNode = iconType ? <Icon type={iconType} /> : null;
    const kids = (children || children === 0)
      ? React.Children.map(children, child => insertSpace(child, this.isNeedInserted())) : null;

    if ('href' in rest) {
      return (
        <a
          {...rest}
          className={classes}
          onClick={this.handleClick}
        >
        {iconNode}{kids}
        </a>
      );
    } else {
      // React does not recognize the `htmlType` prop on a DOM element. Here we pick it out of `rest`.
      const { htmlType, ...otherProps } = rest;

      return (
        <button
          {...otherProps}
          type={htmlType || 'button'}
          className={classes}
          onClick={this.handleClick}
        >
        {iconNode}{kids}
        </button>
      );
    }
  }
}
