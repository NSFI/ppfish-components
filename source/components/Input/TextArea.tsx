import * as React from 'react';
import omit from 'omit.js';
import classNames from 'classnames';
import calculateNodeHeight from './calculateNodeHeight';

function onNextFrame(cb: () => void) {
  if (window.requestAnimationFrame) {
    return window.requestAnimationFrame(cb);
  }
  return window.setTimeout(cb, 1);
}

function clearNextFrameAction(nextFrameId: number) {
  if (window.cancelAnimationFrame) {
    window.cancelAnimationFrame(nextFrameId);
  } else {
    window.clearTimeout(nextFrameId);
  }
}

export interface AutoSizeType {
  minRows?: number;
  maxRows?: number;
}

export type HTMLTextareaProps = React.TextareaHTMLAttributes<HTMLTextAreaElement>;

export interface TextAreaProps extends HTMLTextareaProps {
  prefixCls?: string;
  autosize?: boolean | AutoSizeType;
  onPressEnter?: React.KeyboardEventHandler<HTMLTextAreaElement>;
}

export interface TextAreaRef {
  textAreaRef: HTMLTextAreaElement;
  focus: () => void;
  blur: () => void;
  resizeTextarea: () => void;
}

const InternalTextArea: React.ForwardRefRenderFunction<TextAreaRef, TextAreaProps> = (
  props,
  ref,
) => {
  const [textareaStyles, setTextareaStyles] = React.useState<React.CSSProperties>({});
  const nextFrameActionIdRef = React.useRef<number>(null);
  const textAreaRef = React.useRef<HTMLTextAreaElement>(null);

  const resizeTextarea = () => {
    const { autosize } = props;
    const textAreaNode = textAreaRef.current;
    if (!autosize || !textAreaNode) {
      return;
    }
    const minRows = autosize ? (autosize as AutoSizeType).minRows : null;
    const maxRows = autosize ? (autosize as AutoSizeType).maxRows : null;
    const textareaStyles = calculateNodeHeight(textAreaNode, false, minRows, maxRows);
    setTextareaStyles(textareaStyles);
  };

  const getTextAreaClassName = () => {
    const { prefixCls, className, disabled } = props;
    return classNames(prefixCls, className, {
      [`${prefixCls}-disabled`]: disabled,
    });
  };

  const handleTextareaChange = (e: React.ChangeEvent<HTMLTextAreaElement>) => {
    if (!('value' in props)) {
      resizeTextarea();
    }
    const { onChange } = props;
    if (onChange) {
      onChange(e);
    }
  };

  const handleKeyDown = (e: React.KeyboardEvent<HTMLTextAreaElement>) => {
    const { onPressEnter, onKeyDown } = props;
    if (e.keyCode === 13 && onPressEnter) {
      onPressEnter(e);
    }
    if (onKeyDown) {
      onKeyDown(e);
    }
  };

  React.useImperativeHandle(ref, () => ({
    textAreaRef: textAreaRef.current,
    focus: () => {
      textAreaRef.current.focus();
    },
    blur: () => {
      textAreaRef.current.blur();
    },
    resizeTextarea,
  }));

  React.useEffect(() => {
    const nextFrameActionId = nextFrameActionIdRef.current;
    if (nextFrameActionId) {
      clearNextFrameAction(nextFrameActionId);
    }
    nextFrameActionIdRef.current = onNextFrame(resizeTextarea);
  }, [props.value]);

  const otherProps = omit(props, ['prefixCls', 'onPressEnter', 'autosize']);
  const style = {
    ...props.style,
    ...textareaStyles,
  };
  // Fix https://github.com/ant-design/ant-design/issues/6776
  // Make sure it could be reset when using form.getFieldDecorator
  if ('value' in otherProps) {
    otherProps.value = otherProps.value || '';
  }
  return (
    <textarea
      {...otherProps}
      className={getTextAreaClassName()}
      style={style}
      onKeyDown={handleKeyDown}
      onChange={handleTextareaChange}
      ref={textAreaRef}
    />
  );
};

const TextArea = React.forwardRef(InternalTextArea);

TextArea.displayName = 'TextArea';

TextArea.defaultProps = {
  prefixCls: 'fishd-input',
};

export default TextArea;
