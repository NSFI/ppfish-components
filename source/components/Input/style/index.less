@import "../../../assets/css/index";
@import "./mixin";

// Input styles
.@{fishd-prefix}-input {
  .reset-component;
  .input;
}

//== Style for input-group: input with label, with button or dropdown...
.@{fishd-prefix}-input-group {
  .reset-component;
  .input-group(~"@{fishd-prefix}-input");
  &-wrapper {
    display: inline-block;
    vertical-align: top; // https://github.com/ant-design/ant-design/issues/6403
    width: 100%;
  }
}

// Input with affix: prefix or suffix
.@{fishd-prefix}-input-affix-wrapper {
  .reset-component;
  .input-affix-wrapper(~"@{fishd-prefix}-input");

  // https://github.com/ant-design/ant-design/issues/6144
  .@{fishd-prefix}-input {
    min-height: 100%;  // use min-height, assume that no smaller height to override
  }
}

@import "./search-input";
