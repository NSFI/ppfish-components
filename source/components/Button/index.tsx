import Button from './Button';
import ButtonGroup from './ButtonGroup';

export { ButtonProps, ButtonShape, ButtonSize, ButtonType } from './Button';
export { ButtonGroupProps } from './ButtonGroup';

Button.Group = ButtonGroup;
export default Button;
