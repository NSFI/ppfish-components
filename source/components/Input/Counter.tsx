import * as React from 'react';
import omit from 'omit.js';
import classNames from 'classnames';
import TextArea, { TextAreaProps, TextAreaRef } from './TextArea';
import useControlledState from '../../hooks/useControlledState';

function countValue(value: string) {
  return value.length;
}

export interface CounterProps extends TextAreaProps {
  inputPrefixCls?: string;
  prefixCls?: string;
  limit: number;
  count?: (value: string) => number;
  value?: any;
  defaultValue?: any;
}

export interface CounterRef {
  focus: () => void;
  blur: () => void;
  textarea: TextAreaRef;
}

const InternalCounter: React.ForwardRefRenderFunction<CounterRef, CounterProps> = (props, ref) => {
  const [value, setValue] = useControlledState('', {
    value: props.value,
    defaultValue: props.defaultValue,
  });
  const textareaRef = React.useRef<TextAreaRef>(null);

  const focus = () => {
    textareaRef.current.focus();
  };

  const handleClick = () => {
    focus();
  };

  const handleTextareaChange = (e: React.ChangeEvent<HTMLTextAreaElement>) => {
    const { onChange } = props;
    const textarea = textareaRef.current;
    const textareaValue = textarea && textarea.textAreaRef.value;
    setValue(textareaValue);
    if (onChange) {
      onChange(e);
    }
  };

  const getCount = () => {
    const { count } = props;
    if (!value) {
      return 0;
    }
    // 自定义计数方法
    if (count) {
      return count(String(value));
    }
    return countValue(String(value));
  };

  React.useImperativeHandle(ref, () => ({
    focus,
    blur: () => {
      textareaRef.current.blur();
    },
    textarea: textareaRef.current,
  }));

  const { inputPrefixCls, className, prefixCls, disabled, limit } = props;
  const inputClassName = classNames(className, {
    [`${prefixCls}`]: true,
    [`${inputPrefixCls}-disabled`]: disabled,
  });
  const textareaClassName = classNames(inputPrefixCls, className);
  const otherProps = omit(props, [
    'inputPrefixCls',
    'prefixCls',
    'limit',
    'count',
    'value',
    'onChange',
  ]);
  const total = getCount();
  return (
    <span className={inputClassName} onClick={handleClick}>
      <TextArea
        {...otherProps}
        className={textareaClassName}
        maxLength={limit}
        onChange={handleTextareaChange}
        value={value}
        ref={textareaRef}
      />
      <span className={`${prefixCls}-footer`}>
        <span className={`${prefixCls}-indicator`}>
          {total}/{limit}
        </span>
      </span>
    </span>
  );
};

const Counter = React.forwardRef(InternalCounter);

Counter.displayName = 'Counter';

Counter.defaultProps = {
  inputPrefixCls: 'fishd-input',
  prefixCls: 'fishd-input-counter',
};

export default Counter;
