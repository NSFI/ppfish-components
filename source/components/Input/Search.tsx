import * as React from 'react';
import * as classNames from 'classnames';
import Input, { InputProps } from './Input';
import Icon from '../Icon';
import Button from '../Button';

export interface SearchProps extends InputProps {
  inputPrefixCls?: string;
  onSearch?: (value: string, event?: React.MouseEvent<HTMLElement> | React.KeyboardEvent<HTMLInputElement>) => any;
  enterButton?: boolean | React.ReactNode;
}

export default class Search extends React.Component<SearchProps, any> {
  static defaultProps = {
    inputPrefixCls: 'fishd-input',
    prefixCls: 'fishd-input-search',
    enterButton: false,
  };

  private input: Input;

  onSearch = (e: React.MouseEvent<HTMLElement> | React.KeyboardEvent<HTMLInputElement>) => {
    const { onSearch } = this.props;
    if (onSearch) {
      onSearch(this.input.input.value, e);
    }
    this.input.focus();
  }

  focus() {
    this.input.focus();
  }

  blur() {
    this.input.blur();
  }

  saveInput = (node: Input) => {
    this.input = node;
  }

  getButtonOrIcon() {
    const { enterButton, prefixCls, size, disabled } = this.props;
    const enterButtonAsElement = enterButton as React.ReactElement<any>;
    let node;
    if (!enterButton) {
      node = <Icon className={`${prefixCls}-icon`} type="search" key="searchIcon" />;
    } else if (enterButtonAsElement.type === Button || enterButtonAsElement.type === 'button') {
      node = React.cloneElement(enterButtonAsElement, enterButtonAsElement.type === Button ? {
        className: `${prefixCls}-button`,
        size,
      } : {});
    } else {
      node = (
        <Button
          className={`${prefixCls}-button`}
          type="primary"
          size={size}
          disabled={disabled}
          key="enterButton"
        >
          {enterButton === true ? <Icon type="search" /> : enterButton}
        </Button>
      );
    }
    return React.cloneElement(node, {
      onClick: this.onSearch,
    });
  }

  render() {
    const { className, prefixCls, inputPrefixCls, size, suffix,  enterButton, ...others } = this.props;
    delete (others as any).onSearch;
    const buttonOrIcon = this.getButtonOrIcon();
    const searchSuffix = suffix ? [suffix, buttonOrIcon] : buttonOrIcon;
    const inputClassName = classNames(prefixCls, className, {
      [`${prefixCls}-enter-button`]: !!enterButton,
      [`${prefixCls}-${size}`]: !!size,
    });
    return (
      <Input
        onPressEnter={this.onSearch}
        {...others}
        size={size}
        className={inputClassName}
        prefixCls={inputPrefixCls}
        suffix={searchSuffix}
        ref={this.saveInput}
      />
    );
  }
}
