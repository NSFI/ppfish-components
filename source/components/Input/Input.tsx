import * as React from 'react';
import * as PropTypes from 'prop-types';
import classNames from 'classnames';
import omit from 'omit.js';
import Group from './Group';
import Search from './Search';
import TextArea from './TextArea';
import Counter from './Counter';
import { Omit } from '../../utils/type';

function fixControlledValue<T>(value: T) {
  if (typeof value === 'undefined' || value === null) {
    return '';
  }
  return value;
}

export interface InputProps
  extends Omit<React.InputHTMLAttributes<HTMLInputElement>, 'size' | 'prefix'> {
  prefixCls?: string;
  className?: string;
  size?: 'large' | 'default' | 'small';
  onPressEnter?: React.KeyboardEventHandler<HTMLInputElement>;
  addonBefore?: React.ReactNode;
  addonAfter?: React.ReactNode;
  prefix?: React.ReactNode;
  suffix?: React.ReactNode;
}

export interface InputRef {
  input: HTMLInputElement;
  focus: () => void;
  blur: () => void;
}

const InternalInput: React.ForwardRefRenderFunction<InputRef, InputProps> = (props, ref) => {
  const inputRef = React.useRef(null);
  React.useImperativeHandle(ref, () => ({
    input: inputRef.current,
    focus: () => {
      inputRef.current.focus();
    },
    blur: () => {
      inputRef.current.blur();
    },
  }));

  const handleKeyDown = (e: React.KeyboardEvent<HTMLInputElement>) => {
    const { onPressEnter, onKeyDown } = props;
    if (e.keyCode === 13 && onPressEnter) {
      onPressEnter(e);
    }
    if (onKeyDown) {
      onKeyDown(e);
    }
  };

  const getInputClassName = () => {
    const { prefixCls, size, disabled } = props;
    return classNames(prefixCls, {
      [`${prefixCls}-sm`]: size === 'small',
      [`${prefixCls}-lg`]: size === 'large',
      [`${prefixCls}-disabled`]: disabled,
    });
  };

  const renderLabeledInput = (children: React.ReactElement<any>) => {
    // Not wrap when there is not addons
    if (!props.addonBefore && !props.addonAfter) {
      return children;
    }

    const wrapperClassName = `${props.prefixCls}-group`;
    const addonClassName = `${wrapperClassName}-addon`;
    const addonBefore = props.addonBefore ? (
      <span className={addonClassName}>{props.addonBefore}</span>
    ) : null;

    const addonAfter = props.addonAfter ? (
      <span className={addonClassName}>{props.addonAfter}</span>
    ) : null;

    const className = classNames(`${props.prefixCls}-wrapper`, {
      [wrapperClassName]: addonBefore || addonAfter,
    });

    const groupClassName = classNames(`${props.prefixCls}-group-wrapper`, {
      [`${props.prefixCls}-group-wrapper-sm`]: props.size === 'small',
      [`${props.prefixCls}-group-wrapper-lg`]: props.size === 'large',
    });

    // Need another wrapper for changing display:table to display:inline-block
    // and put style prop in wrapper
    if (addonBefore || addonAfter) {
      return (
        <span className={groupClassName} style={props.style}>
          <span className={className}>
            {addonBefore}
            {React.cloneElement(children, { style: null })}
            {addonAfter}
          </span>
        </span>
      );
    }
    return (
      <span className={className}>
        {addonBefore}
        {children}
        {addonAfter}
      </span>
    );
  };

  const renderLabeledIcon = (children: React.ReactElement<any>) => {
    if (!('prefix' in props || 'suffix' in props)) {
      return children;
    }

    const prefix = props.prefix ? (
      <span className={`${props.prefixCls}-prefix`}>{props.prefix}</span>
    ) : null;

    const suffix = props.suffix ? (
      <span className={`${props.prefixCls}-suffix`}>{props.suffix}</span>
    ) : null;

    const affixWrapperCls = classNames(props.className, `${props.prefixCls}-affix-wrapper`, {
      [`${props.prefixCls}-affix-wrapper-sm`]: props.size === 'small',
      [`${props.prefixCls}-affix-wrapper-lg`]: props.size === 'large',
    });
    return (
      <span className={affixWrapperCls} style={props.style}>
        {prefix}
        {React.cloneElement(children, {
          style: null,
          className: getInputClassName(),
        })}
        {suffix}
      </span>
    );
  };

  const renderInput = () => {
    const { value, className } = props;
    // Fix https://fb.me/react-unknown-prop
    const otherProps = omit(props, [
      'prefixCls',
      'onPressEnter',
      'addonBefore',
      'addonAfter',
      'prefix',
      'suffix',
    ]);

    if ('value' in props) {
      otherProps.value = fixControlledValue(value);
      // Input elements must be either controlled or uncontrolled,
      // specify either the value prop, or the defaultValue prop, but not both.
      delete otherProps.defaultValue;
    }
    return renderLabeledIcon(
      <input
        {...otherProps}
        className={classNames(getInputClassName(), className)}
        onKeyDown={handleKeyDown}
        ref={inputRef}
      />,
    );
  };

  return renderLabeledInput(renderInput());
};

interface InputInterface
  extends React.ForwardRefExoticComponent<InputProps & React.RefAttributes<InputRef>> {
  Group: typeof Group;
  Search: typeof Search;
  TextArea: typeof TextArea;
  Counter: typeof Counter;
}

const Input = React.forwardRef(InternalInput) as InputInterface;

Input.displayName = 'Input';

Input.defaultProps = {
  prefixCls: 'fishd-input',
  type: 'text',
  disabled: false,
};

Input.propTypes = {
  type: PropTypes.string,
  id: PropTypes.oneOfType([PropTypes.string, PropTypes.number]),
  size: PropTypes.oneOf(['small', 'default', 'large']),
  maxLength: PropTypes.oneOfType([PropTypes.string, PropTypes.number]),
  disabled: PropTypes.bool,
  value: PropTypes.any,
  defaultValue: PropTypes.any,
  className: PropTypes.string,
  addonBefore: PropTypes.node,
  addonAfter: PropTypes.node,
  prefixCls: PropTypes.string,
  onPressEnter: PropTypes.func,
  onKeyDown: PropTypes.func,
  onKeyUp: PropTypes.func,
  onFocus: PropTypes.func,
  onBlur: PropTypes.func,
  prefix: PropTypes.node,
  suffix: PropTypes.node,
};

export default Input;
