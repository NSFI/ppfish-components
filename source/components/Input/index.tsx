import Input from './Input';
import Group from './Group';
import Search from './Search';
import TextArea from './TextArea';
import './style/index.less';

export {InputProps} from './Input';
export {GroupProps} from './Group';
export {SearchProps} from './Search';
export {TextAreaProps} from './TextArea';

Input.Group = Group;
Input.Search = Search;
Input.TextArea = TextArea;
export default Input;
