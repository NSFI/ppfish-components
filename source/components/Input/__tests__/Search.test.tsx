import React from 'react';
import { mount } from 'enzyme';
import Input from '../index';
import Button from '../../Button';

const Search = Input.Search;

describe('Input.Search', () => {
  it('should support focus() blur(), and can get input element', function () {
    const ref = React.createRef<any>();
    mount(<Input ref={ref} />);
    ref.current.focus();
    ref.current.blur();
    expect(ref.current.input instanceof HTMLInputElement).toBe(true);
  });

  it('should support custom button', () => {
    const wrapper = mount(<Search enterButton={<button type="button">ok</button>} />);
    expect(wrapper.render()).toMatchSnapshot();
  });

  it('should support custom Button', () => {
    const wrapper = mount(<Search enterButton={<Button>ok</Button>} />);
    expect(wrapper.render()).toMatchSnapshot();
  });

  it('should support enterButton null', () => {
    expect(() => {
      mount(<Search enterButton={null} />);
    }).not.toThrow();
  });

  it('should support ReactNode suffix without error', () => {
    const wrapper = mount(<Search suffix={<div>ok</div>} />);
    expect(wrapper.render()).toMatchSnapshot();
  });

  it('should disable enter button when disabled prop is true', () => {
    const wrapper = mount(<Search placeholder="input search text" enterButton disabled />);
    expect(wrapper.find('.fishd-btn-primary[disabled]')).toHaveLength(1);
  });

  it('should disable search icon when disabled prop is true', () => {
    const onSearch = jest.fn();
    const wrapper = mount(<Search defaultValue="search text" onSearch={onSearch} disabled />);
    wrapper.find('.fishd-input-search-icon').at(0).simulate('click');
    expect(onSearch).toHaveBeenCalledTimes(0);
  });

  it('should trigger onSearch when click search icon', () => {
    const onSearch = jest.fn();
    const wrapper = mount(<Search defaultValue="search text" onSearch={onSearch} />);
    wrapper.find('.fishd-input-search-icon').at(0).simulate('click');
    expect(onSearch).toHaveBeenCalledTimes(1);
    expect(onSearch).toHaveBeenCalledWith(
      'search text',
      expect.objectContaining({
        type: 'click',
        preventDefault: expect.any(Function),
      }),
    );
  });

  it('should trigger onSearch when click search button', () => {
    const onSearch = jest.fn();
    const wrapper = mount(<Search defaultValue="search text" enterButton onSearch={onSearch} />);
    wrapper.find('.fishd-input-search-button').at(0).simulate('click');
    expect(onSearch).toHaveBeenCalledTimes(1);
    expect(onSearch).toHaveBeenCalledWith(
      'search text',
      expect.objectContaining({
        type: 'click',
        preventDefault: expect.any(Function),
      }),
    );
  });

  it('should trigger onSearch when click search button with text', () => {
    const onSearch = jest.fn();
    const wrapper = mount(
      <Search defaultValue="search text" enterButton="button text" onSearch={onSearch} />,
    );
    wrapper.find('.fishd-input-search-button').at(0).simulate('click');
    expect(onSearch).toHaveBeenCalledTimes(1);
    expect(onSearch).toHaveBeenCalledWith(
      'search text',
      expect.objectContaining({
        type: 'click',
        preventDefault: expect.any(Function),
      }),
    );
  });

  it('should trigger onSearch when click search button with customize button', () => {
    const onSearch = jest.fn();
    const wrapper = mount(
      <Search
        defaultValue="search text"
        enterButton={<Button>fishd button</Button>}
        onSearch={onSearch}
      />,
    );
    wrapper.find('Button').at(0).simulate('click');
    expect(onSearch).toHaveBeenCalledTimes(1);
    expect(onSearch).toHaveBeenCalledWith(
      'search text',
      expect.objectContaining({
        type: 'click',
        preventDefault: expect.any(Function),
      }),
    );
  });

  it('should trigger onSearch when click search button of native', () => {
    const onSearch = jest.fn();
    const wrapper = mount(
      <Search
        defaultValue="search text"
        enterButton={<button type="button">native button</button>}
        onSearch={onSearch}
      />,
    );
    wrapper.find('button').simulate('click');
    expect(onSearch).toHaveBeenCalledTimes(1);
    expect(onSearch).toHaveBeenCalledWith(
      'search text',
      expect.objectContaining({
        type: 'click',
        preventDefault: expect.any(Function),
      }),
    );
  });

  it('should trigger onSearch when press enter', () => {
    const onSearch = jest.fn();
    const wrapper = mount(<Search defaultValue="search text" onSearch={onSearch} />);
    wrapper.find('input').simulate('keydown', { key: 'Enter', keyCode: 13 });
    expect(onSearch).toHaveBeenCalledTimes(1);
    expect(onSearch).toHaveBeenCalledWith(
      'search text',
      expect.objectContaining({
        type: 'keydown',
        preventDefault: expect.any(Function),
      }),
    );
  });

  it('should support addonAfter', () => {
    const addonAfter = <span>Addon After</span>;
    const wrapper = mount(<Search addonAfter={addonAfter} />);
    const wrapperWithEnterButton = mount(<Search enterButton addonAfter={addonAfter} />);
    expect(wrapper.render()).toMatchSnapshot();
    expect(wrapperWithEnterButton.render()).toMatchSnapshot();
  });

  it('should support invalid suffix', () => {
    const wrapper = mount(<Search suffix={[]} />);
    expect(wrapper.render()).toMatchSnapshot();
  });

  it('should support invalid addonAfter', () => {
    const wrapper = mount(<Search addonAfter={[]} enterButton />);
    expect(wrapper.render()).toMatchSnapshot();
  });

  it('not crash when use function ref', () => {
    const ref = jest.fn();
    const wrapper = mount(<Search ref={ref} enterButton />);
    expect(() => {
      wrapper.find('button').simulate('mousedown');
    }).not.toThrow();
  });
});
