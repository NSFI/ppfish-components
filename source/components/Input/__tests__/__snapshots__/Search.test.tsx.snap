// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`Input.Search should support ReactNode suffix without error 1`] = `
<span
  class="fishd-input-search fishd-input-affix-wrapper"
>
  <input
    class="fishd-input"
    type="text"
  />
  <span
    class="fishd-input-suffix"
  >
    <div>
      ok
    </div>
    <i
      class="fishdicon fishdicon-search-line fishd-input-search-icon"
    />
  </span>
</span>
`;

exports[`Input.Search should support addonAfter 1`] = `
<span
  class="fishd-input-group-wrapper"
>
  <span
    class="fishd-input-wrapper fishd-input-group"
  >
    <span
      class="fishd-input-search fishd-input-affix-wrapper"
    >
      <input
        class="fishd-input"
        type="text"
      />
      <span
        class="fishd-input-suffix"
      >
        <i
          class="fishdicon fishdicon-search-line fishd-input-search-icon"
        />
      </span>
    </span>
    <span
      class="fishd-input-group-addon"
    >
      <span>
        Addon After
      </span>
    </span>
  </span>
</span>
`;

exports[`Input.Search should support addonAfter 2`] = `
<span
  class="fishd-input-group-wrapper"
>
  <span
    class="fishd-input-wrapper fishd-input-group"
  >
    <span
      class="fishd-input-search fishd-input-search-enter-button fishd-input-affix-wrapper"
    >
      <input
        class="fishd-input"
        type="text"
      />
      <span
        class="fishd-input-suffix"
      >
        <button
          class="fishd-btn fishd-input-search-button fishd-btn-primary"
          type="button"
        >
          <i
            class="fishdicon fishdicon-search-line"
          />
        </button>
      </span>
    </span>
    <span
      class="fishd-input-group-addon"
    >
      <span>
        Addon After
      </span>
    </span>
  </span>
</span>
`;

exports[`Input.Search should support custom Button 1`] = `
<span
  class="fishd-input-search fishd-input-search-enter-button fishd-input-affix-wrapper"
>
  <input
    class="fishd-input"
    type="text"
  />
  <span
    class="fishd-input-suffix"
  >
    <button
      class="fishd-btn fishd-input-search-button"
      type="button"
    >
      <span>
        ok
      </span>
    </button>
  </span>
</span>
`;

exports[`Input.Search should support custom button 1`] = `
<span
  class="fishd-input-search fishd-input-search-enter-button fishd-input-affix-wrapper"
>
  <input
    class="fishd-input"
    type="text"
  />
  <span
    class="fishd-input-suffix"
  >
    <button
      type="button"
    >
      ok
    </button>
  </span>
</span>
`;

exports[`Input.Search should support invalid addonAfter 1`] = `
<span
  class="fishd-input-group-wrapper"
>
  <span
    class="fishd-input-wrapper fishd-input-group"
  >
    <span
      class="fishd-input-search fishd-input-search-enter-button fishd-input-affix-wrapper"
    >
      <input
        class="fishd-input"
        type="text"
      />
      <span
        class="fishd-input-suffix"
      >
        <button
          class="fishd-btn fishd-input-search-button fishd-btn-primary"
          type="button"
        >
          <i
            class="fishdicon fishdicon-search-line"
          />
        </button>
      </span>
    </span>
    <span
      class="fishd-input-group-addon"
    />
  </span>
</span>
`;

exports[`Input.Search should support invalid suffix 1`] = `
<span
  class="fishd-input-search fishd-input-affix-wrapper"
>
  <input
    class="fishd-input"
    type="text"
  />
  <span
    class="fishd-input-suffix"
  >
    <i
      class="fishdicon fishdicon-search-line fishd-input-search-icon"
    />
  </span>
</span>
`;
